Aiida1 cleanmerge

danieleongari opened the pull request:
This is a merge of the two branches that contains:
- advanced parsing
- multistage WorkChain with default protocols and tests

Also it changes some settings:
- ~~parameters  `required=False` for the input, to make the design of WorkChains simpler, exposing everything without excluding parameters~~ *Edit: now using `cp2k_base.cp2k.parameters` input.*
- retrieving the .dcd files of the trajectory (that are usually lighter than the .out file)

~~It requires a fix: now the choice of the parser is hard-coded (https://github.com/danieleongari/aiida-cp2k/blob/aiida1_merge/aiida_cp2k/parsers/__init__.py#L57) and it needs to be selected with the metadata of the calculation. This "Advanced" parsing is robust and works with the previous tests, but to be safer we would like to switch it on by the expert user (or the specific workchain). @ltalirz suggested to create a separate class (e.g., `Cp2kAdvancedParser(Parser)`), but this solution looks equally fine to me, if it is possible to pass the variable ` parser_type` to `Cp2kParser`~~ *Edit: problem solved with `cp2k.metadata.options.parser_name = cp2k_advanced_parser`*

danieleongari commented:
yapf probably failed because I changed the style from `google` to `pep8`. 
Indeed with the previous one, some short dictionaries like:
```
    inputs = {
        'structure': structure,
        'parameters': parameters,
        'code': code,
        'metadata': {
            'options': options,
        }
    }
```
Were condensed to a single line (being shorter than the threshold of 120 characters)

yakutovicha commented on examples/workchains/test_multistage_Al.py:
```diff
@@ -0,0 +1,81 @@
+        }
+    }
+
+    run(Cp2kMultistageWorkChain, **inputs)
```
@danieleongari can you please add some checks here that the output parameters have the expected value? Same for the other tests.

danieleongari replied:
Not sure this is safe, because the Multistage can terminate in very different ways

yakutovicha commented on examples/workchains/test_multistage_h2o_testfile.py:
```diff
@@ -0,0 +1,57 @@
+        "max_wallclock_seconds": 1 * 3 * 60,
+    }
+    inputs = {
+        'protocol_yaml': SinglefileData(file=os.path.abspath(os.path.join(thisdir, '../data/testfile.yaml'))),
```
```suggestion
        'protocol_yaml': SinglefileData(file=os.path.abspath(os.path.join(thisdir, '..', 'data', 'testfile.yaml'))),
```

yakutovicha commented on aiida_cp2k/workchains/multistage.py:
```diff
@@ -0,0 +1,486 @@
+HARTREE2EV = 27.211399
+
+
+def merge_dict(dct, merge_dct):
```
Please move it into the `utils`folder

danieleongari replied:
See next

yakutovicha commented on aiida_cp2k/workchains/multistage.py:
```diff
@@ -0,0 +1,486 @@
+    return Dict(dict=p1_dict).store()
+
+
+def get_kinds_section(structure, protocol_settings):
```
Please move it into the `utils` folder. Also consider whether to move into `utils` the functions below.

danieleongari replied:
I propose to move all the functions to utils, except for `extract_results` that is very much tailor-made for the multistage workchain and it makes more sense to include it in the same file

yakutovicha commented on aiida_cp2k/workchains/multistage.py:
```diff
@@ -0,0 +1,486 @@
+            self.ctx.parameters_yaml = yaml.safe_load(self.inputs.protocol_yaml.open())
+        else:
+            thisdir = os.path.dirname(os.path.abspath(__file__))
+            yamlfullpath = '{}/multistage_protocols/{}.yaml'.format(thisdir, self.inputs.protocol_tag.value)
```
```suggestion
            yamlfullpath = os.path.join(thisdir, 'multistage_protocols',  self.inputs.protocol_tag.value+'.yaml')
```

yakutovicha replied:
Please check if I overlooked similar things in the other files

yakutovicha commented on aiida_cp2k/workchains/multistage.py:
```diff
@@ -0,0 +1,486 @@
+        # Inputs
+        spec.expose_inputs(Cp2kBaseWorkChain,
+                           namespace='cp2k_base',
+                           exclude=['parameters', 'cp2k.metadata.options.parser_name'])
```
```suggestion
                           exclude=['cp2k.parameters', 'cp2k.metadata.options.parser_name'])
```
I think it should be 'cp2k.parameters', right?

danieleongari replied:
Not sure, we need to look carefully together with the first comment on how to make the input parameter optional.

yakutovicha replied:
I think if you exclude `cp2k.parameters`, then the input parameter (from the first comment) can remain `required=True`.

yakutovicha replied:
If I didn't overlook anything, in the multistage work chain you anyways generate this input parameter on the fly

danieleongari replied:
In the test https://github.com/danieleongari/aiida-cp2k/blob/aiida1_cleanmerge/examples/workchains/test_multistage_h2o_fail.py
I do modify the parameters, and I like to keep the input like this so that an advanced user can always modify the final CP2K setting independently  from the protocol.

danieleongari replied:
The best here would be to create an input that is to use your suggestion and then have a `base_cp2k.cp2k.parameters` that is not required, so that we can keep `required=True` in the Cp2kCalc. However, I don't know if this is possible, i.e., to make a "nested" input.

danieleongari replied:
I saw this is possible:
```
        spec.expose_inputs(Cp2kBaseWorkChain,
                           namespace='cp2k_base',
                           exclude=['cp2k.parameters', 'cp2k.metadata.options.parser_name'])

        spec.input('cp2k_base.cp2k.parameters',
                   valid_type=Dict,
                   required=False,
                   help='Specify custom CP2K settings that will be overwritten just before submitting the calculation',
                   )
```

yakutovicha commented on aiida_cp2k/utils/parser.py:
```diff
@@ -43,6 +43,235 @@ def parse_cp2k_output(fobj):
     return result_dict
 
 
+def parse_cp2k_output_advanced(fobj):  # pylint: disable=too-many-locals, too-many-statements, too-many-branches
```
Can this function be splitted into smaller sub-parts? Also, could you please add a couple of pytests for that?

danieleongari replied:
I don't think I can, since they are a list of `if` instances on the line which is read. I can take out some parts but this would make the readability much more complicated. I propose to keep it  like this since it is still quite experimental.

danieleongari commented on aiida_cp2k/calculations/__init__.py:
```diff
@@ -30,16 +30,17 @@ class Cp2kCalculation(CalcJob):
 
         # Input parameters
-        spec.input('parameters', valid_type=Dict, help='the input parameters')
+        spec.input('parameters', valid_type=Dict, required=False, help='the input parameters')
```
This is because you can expose all the inputs from the CP2K calculation in a workchain that creates the input parameters internally, but you still have the option of forcing certain parameters of the calculation. One could exclude this input parameters in the workchain and create another input that is required=False, but it is a more elaborate option.

yakutovicha replied:
Hmm, I think it would would work anyways as long as you don't expose the input `cp2k.parameters` in the multistage work chain. And, as far as I can see, you do not expose it.

danieleongari replied:
See the later comment on `expose input`

danieleongari replied:
Agree on  not changing it

yakutovicha commented:
BTW, I completely forgot: use should use `calcfunction` instead of `workfunction`. In principle the `work function`is not permitted to generate new data objects.

danieleongari commented:
> BTW, I completely forgot: use should use `calcfunction` instead of `workfunction`. In principle the `work function`is not permitted to generate new data objects.

Can you link a tutorial/manual page that explains clearly this point?

yakutovicha commented:
> > BTW, I completely forgot: use should use `calcfunction` instead of `workfunction`. In principle the `work function`is not permitted to generate new data objects.
> 
> Can you link a tutorial/manual page that explains clearly this point?

sure, have a look here: https://aiida-core.readthedocs.io/en/latest/concepts/processes.html#process-types

danieleongari commented:
I addressed your comments, I now need to:
1) Extensively test the workchain and see if the "default" protocols are valid
2) Check if the overall input/output info of the workchain are ok (i.e., I don't need to add anything essential that is not retro-compatible)
3) Make tests for the parser (I will have a look but I require your help @yakutovicha on how to implement it properly)

danieleongari commented:
@yakutovicha 
Just experienced a problem:
when I call MultistageWC from an upper WC (e.g., https://github.com/danieleongari/aiida-ddec/blob/aiida1/aiida_ddec/workchains/cp2kmultistage_ddec.py) the section 
```
        spec.input('cp2k_base.cp2k.metadata.options.parser_name',
                   valid_type=six.string_types,
                   default='cp2k_advanced_parser',
                   non_db=True,
                   help='Parser of the calculation: the default is cp2k_advanced_parser to get the necessary info')
``` 
is somehow not working and the work fails because `parser_name=cp2k_base_parser`:
```
AttributeDict({'metadata': AttributeDict({'store_provenance': True, 'call_link_label': 'CALL', 'label': 'base(stage_0/settings_0)'}), 'max_iterations': <Int: uuid: 871bdcc5-830d-4472-bd66-52fc26a86ee2 (pk: 18221) value: 5>, 'clean_workdir': <Bool: uuid: 755bd2ae-ee72-411d-a34f-ab53b29d4814 (pk: 18222) value: False>, 'cp2k': AttributeDict({'structure': <StructureData: uuid: f1554c13-3e7d-489b-9bb9-4e73b18313d6 (pk: 18220)>, 'code': <Code: Remote code 'cp2k-5.1' on localhost, pk: 22, uuid: 757a8ff6-1c37-41a3-b094-ed2b800774bc>, 'metadata': AttributeDict({'options': AttributeDict({'resources': AttributeDict({'num_machines': 1}), 'max_wallclock_seconds': 600, 'withmpi': True, 'scheduler_stdout': '_scheduler-stdout.txt', 'scheduler_stderr': '_scheduler-stderr.txt', 'custom_scheduler_commands': '', 'mpirun_extra_params': [], 'import_sys_environment': True, 'environment_variables': AttributeDict({}), 'prepend_text': '', 'append_text': '', 'parser_name': 'cp2k_base_parser'}), 'store_provenance': True, 'call_link_label': 'CALL', 'dry_run': False, 'label': 'GEO_OPT'}), 'file': AttributeDict({}), 'parameters': <Dict: uuid: 32b64032-69e9-469d-b4c6-da73116001a8 (pk: 18232)>})})
```

Why is this happening and how to fix it?

danieleongari commented:
> @yakutovicha
> Just experienced a problem:
> when I call MultistageWC from an upper WC (e.g., https://github.com/danieleongari/aiida-ddec/blob/aiida1/aiida_ddec/workchains/cp2kmultistage_ddec.py) the section
> 
> ```
>         spec.input('cp2k_base.cp2k.metadata.options.parser_name',
>                    valid_type=six.string_types,
>                    default='cp2k_advanced_parser',
>                    non_db=True,
>                    help='Parser of the calculation: the default is cp2k_advanced_parser to get the necessary info')
> ```
> 
> is somehow not working and the work fails because `parser_name=cp2k_base_parser`.
> This is what I obtain if I print `self.ctx.base_inp`:
> 
> ```
> AttributeDict({'metadata': AttributeDict({'store_provenance': True, 'call_link_label': 'CALL', 'label': 'base(stage_0/settings_0)'}), 'max_iterations': <Int: uuid: 871bdcc5-830d-4472-bd66-52fc26a86ee2 (pk: 18221) value: 5>, 'clean_workdir': <Bool: uuid: 755bd2ae-ee72-411d-a34f-ab53b29d4814 (pk: 18222) value: False>, 'cp2k': AttributeDict({'structure': <StructureData: uuid: f1554c13-3e7d-489b-9bb9-4e73b18313d6 (pk: 18220)>, 'code': <Code: Remote code 'cp2k-5.1' on localhost, pk: 22, uuid: 757a8ff6-1c37-41a3-b094-ed2b800774bc>, 'metadata': AttributeDict({'options': AttributeDict({'resources': AttributeDict({'num_machines': 1}), 'max_wallclock_seconds': 600, 'withmpi': True, 'scheduler_stdout': '_scheduler-stdout.txt', 'scheduler_stderr': '_scheduler-stderr.txt', 'custom_scheduler_commands': '', 'mpirun_extra_params': [], 'import_sys_environment': True, 'environment_variables': AttributeDict({}), 'prepend_text': '', 'append_text': '', 'parser_name': 'cp2k_base_parser'}), 'store_provenance': True, 'call_link_label': 'CALL', 'dry_run': False, 'label': 'GEO_OPT'}), 'file': AttributeDict({}), 'parameters': <Dict: uuid: 32b64032-69e9-469d-b4c6-da73116001a8 (pk: 18232)>})})
> ```
> 
> Why is this happening and how to fix it?
> When Multistage is called directly, the change of string works as expected.

SOLVED
I was using 
```
        spec.expose_inputs(Cp2kMultistageWorkChain)
        spec.expose_inputs(Cp2kDdecWorkChain)
```
And since I don't want to use `namespace` that would make the input too nested, the second one was overwriting the first `base_cp2k`. Fixed with:
```
        spec.expose_inputs(Cp2kMultistageWorkChain)
        spec.expose_inputs(Cp2kDdecWorkChain, exclude=['cp2k_base'])
```

yakutovicha commented on aiida_cp2k/workchains/multistage.py:
```diff
@@ -0,0 +1,368 @@
+    # General info
+    output_dict['nstages_valid'] = nstages_valid
+    output_dict['nsettings_discarded'] = nsettings_discarded
+    output_dict['last_tag'] = value.label.split()[-1]
```
This won't always work, cause the items in a dictionary aren't sorted.
